### PR TITLE
I18n support

### LINKED ISSUE #2923: Internationalization (I18N) support for Kiali UI

### PR DESCRIPTION
**Pull request by ferhoyos**
### Describe the change

This PR adds i18n support to Kiali application using `i18next` library and some examples of Chinese translation in different kind of React components (classes, hooks, etc.). Translation of Kiali application in any language is out of scope of this PR.

Translation examples provided are:.
- Navigation menu (external function)
- Home cluster name tooltip in the top navigation menu (React hook)
- Number of labels per namespace (React class)

image

A language dropdown has been created to switch between English and Chinese. New locale value is stored in redux store so the value is persisted on refresh)

image

For the moment supported languages are English (default) and Chinese (in progress).

Translations are stored in json files in `public/locales/${LOCALE}` folder and generated automatically by `i18next-parser` library on yarn command `yarn i18n`. This command is executed automatically on build process to add new statements to translation files (or removed unused ones). English translation file must not be modified since code text is the source of truth for English statements.

I have added two i18n settings in the Kiali server configuration under `kiali_feature_flags/ui_defaults/i18n`:
- language: Default language (if no value is stored in redux store)
- showSelector: If true shows selector language icon

Since Kiali code is the base of OSSMC plugin, I have tried to follow OpenShift dynamic plugin requirements for i18n labels (https://github.com/openshift/console-plugin-template/tree/main?tab=readme-ov-file#i18n):
- Added I18N_NAMESPACE variable as i18n translation namespace (empty for Kiali, proper plugin value for OSSMC).

```
const { t } = useTranslation(I18N_NAMESPACE);
```

- Key should be English value since key separators are not allowed.

```
t('Traffic Graph')
```

### How to translate different React components
**React class**: 
Use `withTranslation` high order component (https://react.i18next.com/latest/withtranslation-hoc)

You have to embed React class with `withTranslation` function (similar to redux `connect` function):
```
export const Menu = withTranslation(I18N_NAMESPACE)(MenuComponent);
```

Besides, `WithTranslation` props must be added to React class properties (like Redux properties)

```
type MenuProps = WithTranslation & {
   ...
};
```

**React hook**
Use `useTranslation` hook (https://react.i18next.com/latest/usetranslation-hook)

```
const { t } = useTranslation(I18N_NAMESPACE);
```

**External function**
Include `t` function as input parameter:

```
const navMenuItems = (t: TFunction): MenuItem[] => {
  return [
    {
      id: 'overview',
      title: t('Overview'),
      to: '/overview',
      pathsActive: [/^\/overview\/(.*)/]
    },


const allNavMenuItems = navMenuItems(this.props.t);
```

**Interpolation**
You can include any variable value within the translated statement with interpolation (https://www.i18next.com/translation-function/interpolation)

```
<div>{t('Kiali home cluster: {{name}}', { name: homeCluster?.name })}</div>
```

The variables are represented with `{{$variable}}`. You can set any name to the variable (in the example `name`), you have just to be sure that matches with json field. 

This is how it looks like in the translation file:
```
"Kiali home cluster: {{name}}": "Kiali主集群: {{name}}",
```

**Plurals**
When a statement is different in singular and plural, you can use defaultValue_one and defaultValue_other fields to add different plural statements in English (no need to add 's' manually anymore). Note that in this case variable name must be `count`. For convenience the key will be the singular statement. More info here: https://www.i18next.com/translation-function/plurals

```
labelsInfo = this.props.t('{{count}} label', {
        count: labelsLength,
        defaultValue_one: '{{count}} label',
        defaultValue_other: '{{count}} labels'
      });
```

**Formatting**
This feature is out of scope of this PR, although it would be interesting to investigate how to format dates for different languages.

### Steps to test the PR
1. Execute `yarn install` to install new i18n libraries
3. Check that switch language dropdown is available
4. Change the language to Chinese and see translated labels

### Automation testing
I have added unit test for `LanguageSwitch` component. For the moment I don't think Cypress test is needed since i18n language switch is disabled by default and Chinese translation is not done yet.

### Issue reference
Closes #2923 

Operator PR: https://github.com/kiali/kiali-operator/pull/746

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -145,7 +150,7 @@
     "redux-mock-store": "1.5.4",
     "sass": "^1.63.6",
     "seedrandom": "3.0.1",
-    "typescript": "4.6.4"
+    "typescript": "^4.9.5"
```
Comment (ferhoyos) [inline frontend/package.json:153]: I have to update Typescript 4 to use latest version of i18n

File: frontend/src/app/AuthenticationController.tsx [hunk 2]
```diff
@@ -218,19 +228,33 @@ class AuthenticationControllerComponent extends React.Component<
...
+  private applyUIDefaults = (): void => {
     const uiDefaults = serverConfig.kialiFeatureFlags.uiDefaults;
+
     if (uiDefaults) {
+      // Set I18n locale
+      let locale = store.getState().globalState.locale;
+
+      // If locale in redux store is empty or not supported, set to default language
+      if (!uiDefaults.i18n.locales.includes(locale)) {
+        locale = uiDefaults.i18n.default ?? Locale.ENGLISH;
```
Comment (jshaughn) [inline frontend/src/app/AuthenticationController.tsx:240]: May want to log a warning message here.

**Review by jshaughn — COMMENTED** [review]
Code looks good, great job with all of this.  I may have done something wrong but I'm not seeing the localized menu items.  I am seeing the dropdown, this is with 'zh' selected:

image

Note that the current locale is not selected in the dropdown.  I think it should be.  Also note, with 'zh' selected the Overview card presents `labels` and with 'en' it presents `label`.

**Comment by ferhoyos** [comment]
@jshaughn that's weird... can you check that you receive translation.json file correctly from backend server?

image

Thanks for the suggestions. I have added a check icon to know which is the selected language:

image


An alternative is to replace switch language icon with typical language dropdown using 3-letter ISO 639 code (https://en.wikipedia.org/wiki/List_of_ISO_639_language_codes)

image

image

**Comment by jshaughn** [comment]
> An alternative is to replace switch language icon with typical language dropdown using 3-letter ISO 639 code

With the latest update I am seeing the check mark.  This is fine for me, no need to change from my side.

> can you check that you receive translation.json file correctly from backend server?

I noticed that things were workibg with `yarn start`, because the proxy URL is defined by me and being pulled from package.json.  The issue is the standard build is using an incorrect URL to pull the translation, and getting a 404.

Failing URL: `http://localhost:20001/locales/en/translation.json`

Working URL: `http://localhost:20001/kiali/locales/en/translation.json`

**Comment by ferhoyos** [comment]
> With the latest update I am seeing the check mark. This is fine for me, no need to change from my side.

Cool, I also like more the language switch icon. Maybe @fkargbo can help us here.

Thanks for catching the proxy issue @jshaughn, I will take a look.

**Review by jshaughn — APPROVED** [review]
Now working in prod mode, excellent work all around, @ferhoyos!

**Comment by fkargbo** [comment]
> > With the latest update I am seeing the check mark. This is fine for me, no need to change from my side.
> 
> Cool, I also like more the language switch icon. Maybe @fkargbo can help us here.
> 
> Thanks for catching the proxy issue @jshaughn, I will take a look.

The language icon looks good to me. It's standard now where most people know what it means. Alternatively, we can add an "on hover" tooltip that reads "Language" or "Switch language" if a user hovers over it.

**Comment by jmazzitelli** [comment]
This new feature warrants a new doc page somewhere on kiali.io. Maybe under "Features"? "Features > I18N support" ??

Do we have plans for kiali.io docs to be merged when this feature is merged?

File: config/config.go [hunk 3]
```diff
@@ -467,6 +467,12 @@ type GraphUIDefaults struct {
 	Traffic     GraphTraffic      `yaml:"traffic,omitempty" json:"traffic,omitempty"`
 }
 
+// I18nUIDefaults defines UI Defaults specific to the I18n settings
+type I18nUIDefaults struct {
+	Default string   `yaml:"default,omitempty" json:"default,omitempty"`
+	Locales []string `yaml:"locales,omitempty" json:"locales,omitempty"`
```
Comment (jshaughn) [inline config/config.go:473]: `Languages` ?

**Comment by jshaughn** [comment]
> This new feature warrants a new doc page somewhere on kiali.io. Maybe under "Features"? "Features > I18N support" ??

We do need docs.  `Features > I18N` sounds like a good location, because it is a feature but also spans configuration and contribution.

**Comment by ferhoyos** [comment]
> > > With the latest update I am seeing the check mark. This is fine for me, no need to change from my side.
> > 
> > 
> > Cool, I also like more the language switch icon. Maybe @fkargbo can help us here.
> > Thanks for catching the proxy issue @jshaughn, I will take a look.
> 
> The language icon looks good to me. It's standard now where most people know what it means. Alternatively, we can add an "on hover" tooltip that reads "Language" or "Switch language" if a user hovers over it.

Thanks for the feedback @fkargbo! Tooltip added to the icon (left position so it does not interfere with dropdown list)

image

**Comment by ferhoyos** [comment]
> This new feature warrants a new doc page somewhere on kiali.io. Maybe under "Features"? "Features > I18N support" ??
> 
> Do we have plans for kiali.io docs to be merged when this feature is merged?

Yes, I will document this feature on kiali.io docs. I will also add i18n section in frontend development README to explain how to translate any label within Kiali application

File: frontend/src/config/ServerConfig.ts [hunk 4]
```diff
@@ -114,8 +114,8 @@ const defaultServerConfig: ComputedServerConfig = {
         }
       },
       i18n: {
-        default: 'en',
-        languages: ['en']
+        language: 'en',
+        showSelector: true
```
Comment (jshaughn) [inline frontend/src/config/ServerConfig.ts:118]: also,  I think`false`.

File: config/config.go [hunk 5]
```diff
@@ -825,8 +825,8 @@ func NewConfig() (c *Config) {
 					},
 				},
 				I18n: I18nUIDefaults{
-					Default:   "en",
-					Languages: []string{"en"},
+					Language:     "en",
+					ShowSelector: true,
```
Comment (jshaughn) [inline config/config.go:829]: I think the default should be `ShowSelector: false`, right?   We don't want to show the selector when most users will just want English.  Only Chinese users will maybe want to see the selector.

**Review by jshaughn — APPROVED** [review]
Looks good!  I wasn't sure if we actually wanted `frontend/public/locales/en/translation.json` in the repo, but I guess it doesn't hurt?

**Comment by ferhoyos** [comment]
It does not hurt unless any malicious PR change It manually (it is the source of truth for i18n). Maybe it is safer to ignore it (I can check it in another PR - i18n dev docs update is pending)

**Pull request merged**